### PR TITLE
[SPARK-15345][SQL][PYSPARK]. SparkSession's conf doesn't take effect when this already an existing SparkContext

### PR DESCRIPTION
**Pull request by zjffdu**
## What changes were proposed in this pull request?

Override the existing SparkContext is the provided SparkConf is different. PySpark part hasn't been fixed yet, will do that after the first round of review to ensure this is the correct approach. 
## How was this patch tested?

Manually verify it in spark-shell. 

@rxin  Please help review it, I think this is a very critical issue for spark 2.0

### CODE REVIEW

**Comment by rxin** [comment]
hm I don't think it is safe to create two contexts and stop the existing one. Maybe we can go ahead and change the sql conf setting, but not the spark context setting?

**Comment by zjffdu** [comment]
Changing sql conf setting seems a better approach. Besides that some other setting in SparkSession#Build also won't take effect if there's already an existing SparkContext. e.g. if there's an local mode SparkContext, then user can not create SparkSession of yarn-cluster mode, At least we need to add warning message for that case.

**Comment by rxin** [comment]
logging a warning seems like a good idea!

**Comment by zjffdu** [comment]
It seems changing sqlconf won't work.  Because SharedState is coupled with SparkContext closely while SparkContext.conf is immutable. That means if SparkContext is created manually before SparkSession, then every setting can not be changed. It looks it would involve more changes to fix it. 

```
/**
 * A class that holds all state shared across sessions in a given [[SQLContext]].
 */
private[sql] class SharedState(val sparkContext: SparkContext) {

  /**
   * Class for caching query results reused in future executions.
   */
  val cacheManager: CacheManager = new CacheManager

  /**
   * A listener for SQL-specific [[org.apache.spark.scheduler.SparkListenerEvent]]s.
   */
  val listener: SQLListener = SQLContext.createListenerAndUI(sparkContext)

```

**Comment by rxin** [comment]
Can you explain more why is it a problem? I think it is OK to not change the undelying SparkContext, since the configs are supposed to be mutable, but only change the SparkSession's configs.

**Comment by zjffdu** [comment]
SharedState is constructed from SparkContext, so if there's already an existing SparkContext, I can not pass additional session conf to SharedState since the conf of SparkContext is immutable. Please correct me if I don't understand it correctly.

**Comment by rxin** [comment]
I've done it here: https://github.com/apache/spark/pull/13200

**Comment by rxin** [comment]
@zjffdu did https://github.com/apache/spark/pull/13200 solve the problem?

**Comment by zjffdu** [comment]
@rxin Seems it didn't resolve the issue described in the jira, I don't have time to check the changes, may take a look at it tomorrow. Another issue I found is that hive-site.xml is not picked up by spark, I see the following log even I put hive-site.xml on classpath ( it should connect to hive metastore service and warehouse path should be a hdfs path)
```
16/05/23 20:45:00 INFO HiveSharedState: Setting Hive metastore warehouse path to '/Users/jzhang/github/spark/spark-warehouse'
```

**Comment by zjffdu** [comment]
test it again.

**Comment by rxin** [comment]
cc @andrewor14

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -2245,6 +2245,7 @@ object SparkContext extends Logging {
       if (activeContext.get() == null) {
         setActiveContext(new SparkContext(config), allowMultipleContexts = false)
       }
+      logWarning("Use an existing SparkContext, some configuration may not take effect.")
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: maybe log this only if there are non-empty configs? same thing for all the warning messages later.

**Comment by zjffdu** [comment]
Thanks @rxin, update the PR.

**Comment by rxin** [comment]
LGTM pending Jenkins.

**Comment by andrewor14** [comment]
LGTM2. Merging into master 2.0

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -771,7 +777,11 @@ object SparkSession {
 
           val sparkConf = new SparkConf()
           options.foreach { case (k, v) => sparkConf.set(k, v) }
-          SparkContext.getOrCreate(sparkConf)
+          val sc = SparkContext.getOrCreate(sparkConf)
+          // maybe this is an existing SparkContext, update its SparkConf which maybe used
```
Comment (tedyu) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: We can check whether the sc was pre-existing, right ?

In that case the foreach below is not needed.

**Pull request closed without merging**